### ISSUE TITLE
Twisted does not have a code of conduct

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8173]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8173|
|Type|enhancement|
|Created|2016-01-12 06:16:55Z|
|Branch|https://github.com/twisted/twisted/tree/explicit-is-better-than-implicit-8173|

It's come up on the mailing list a while ago, everyone was in favour, we just need to push forward and add it.

<details><summary>Searchable metadata</summary>

```
trac-id__8173 8173
type__enhancement enhancement
reporter__hawkowl hawkowl
priority__highest highest
milestone__None None
branch__branches_explicit_is_better_than_implicit_8173 branches/explicit-is-better-than-implicit-8173
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1452579415461137 1452579415461137
changetime__1453349050815588 1453349050815588
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [46611]) Branching to explicit-is-better-than-implicit-8173.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

Okay, I'm happy with this as a starting point. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban commented|
|-|-|

Looks good. Thanks!

Only minor comments.

-----

Why markdown ?

Why not have it as part of the dev policy documentation ? http://twistedmatrix.com/documents/current/core/development/policy/index.html

------

I am not very happy with all this... looks like all the legal disclaimers and legal notes, legal papers and cookies notifications on all EU sites.

I agree that explicit is better than implicit... but ... common sense :)

http://www.i-am-bored.com/2015/12/eu-vs-usa-warning-labels-no-common-sense-require-pic.html

Maybe is also OK to just have a note in the main README like

```
We are committed to making participation in this project a harassment-free
experience for everyone. Please report any deviations.
```

-----------

Was there any harassment case inside the Twisted community? 

---------

Leaving this for review so that someone else can check it.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @hawkowl|
|-|-|

Replying to [adiroiban](#note_3):
> Looks good. Thanks!
> 
> Only minor comments.
> 
> -----
> 
> Why markdown ?
> 
> Why not have it as part of the dev policy documentation ? http://twistedmatrix.com/documents/current/core/development/policy/index.html

Markdown because that's what https://github.com/simonv3/covenant-generator outputs, and we wanted this work to be as minimal as possible.  It may be reasonable to integrate it into our documentation later, as a separate PR.  But in the meanwhile, I would like to just keep up to date as the Contributor Covenant evolves, and update to more recent versions.

> I am not very happy with all this... looks like all the legal disclaimers and legal notes, legal papers and cookies notifications on all EU sites.
> 
> I agree that explicit is better than implicit... but ... common sense :)
> 
> http://www.i-am-bored.com/2015/12/eu-vs-usa-warning-labels-no-common-sense-require-pic.html
> 
> Maybe is also OK to just have a note in the main README like
> 
> ```
> We are committed to making participation in this project a harassment-free
> experience for everyone. Please report any deviations.
> ```

The problem with this suggestion is that it's too vague about what "harassment" is, and also what exactly you should do if you have a problem.  It's unfortunate, but reasonable (although ignorant) people can fail to understand what constitutes harassment, so we have a CoC to spell it out.  More importantly, we specify a way to get in touch if you have a problem.

This discussion has been had over and over again on dozens of open source mailing lists.  We had it on the Twisted mailing list here - https://twistedmatrix.com/pipermail/twisted-python/2015-June/029587.html - and some similar discussions came up.

To address your criticism about "too many rules" specifically, the difference between this and, let's say, cookie notifications, is that cookie notifications are boilerplate that shows up everywhere, disclaimers that are mostly not read and not understood.  This is just in the root of the repository, a notification for anyone who goes looking for it.

If you survey people in underrepresented and marginalized populations in open source and software development - people of color, women, lgbt people and so on - many of them will say that they specifically look for codes of conduct to ensure that the conference or project takes issues like harassment seriously and will try to do something about them.  We are putting this in place as a signal to those folks that we do care, that we do take issues like this seriously, and we will do what we can about them if they occur.

> Was there any harassment case inside the Twisted community? 

Not that we're aware of, but:

 1. There's been plenty of harassment in the broader open source community; [the Geek Feminism Wiki](http://geekfeminism.wikia.com/wiki/Timeline_of_incidents) maintains a list.
 2. We have not had a mechanism for reporting harassment (which this document adds), so it's quite possible some has occurred and we just don't know about it.
 3. Just because you haven't had any murders in a particular town yet doesn't mean you don't want a law against murder there.

> Leaving this for review so that someone else can check it.

I think this is consistent with the community spirit expressed on the mailing list.  Please merge it.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl set status to `closed`|
|-|-|

(In [46635]) Merge explicit-is-better-than-implicit-8173: Adopt a Code of Conduct

Author: hawkowl
Reviewers: adiroiban, glyph
Special Thanks: moshez
Fixes: [#8173](https://github.com/twisted/twisted/issues/8173)
